### DISCUSSION TITLE
Automatic photovoltaic excess charging

### DISCUSSION THREAD

**Discussion by morphZ** [cathiele/homeassistant-goecharger#55] [Ideas]
Hi @cathiele,

I'm using the go-e charger at my home in germany with home assistant (thanks for your integration BTW!)

I'm trying to auto adjust the charge power depending on the excess power from the PV generator on my roof. I wrote something using a finite state machine from `pytransitions` and a pid controller from `simple-pid`. At the moment it's a [seperate integration](https://github.com/morphZ/core/tree/power/homeassistant/components/pvcharge), but I'd feel it could be good to integrate that into `homeassistant-goecharger`.

What do you think? Do you have the same problem? If yes, how do you solve it?

Cheers
morphZ

**Comment by jmtatsch** [comment]
I also have the same interest. 
While I don't think the control logic belongs into the go-echarger module at all, it would be reasonable and tremendously helpful if charging could be controlled by an arbitrary input power instead of amps.
 
This could work like this:

* The solar system calculates an excess power of 4321W 
* go-echarger module has a service  set_charging_power(4321)
* go-echarger module should look in a internal dictionary with possible charging powers which is the closest one e.g. 6*3*230=4140 and activate it

Like that it would be very simple and elegant to just update the charging_power every time a new excess power estimate comes in.

**Comment by klacol** [comment]
I also have the same interest. I carefully read the project [EVCC](https://github.com/evcc-io/evcc) that is recommended by @cathiele in the pinned issue #32. 

The problem for me with EVCC is, that it is a complete new installation on my PI and I dot not know, how to do it parallel to HA.

My Solaredge PV is in HA, therefore it would be nice to have an HA module, that puts the wallbox and the PV modules together.

**Reply by WolfgangSchmid** [reply]
I've created an HA-automation to switch the goe-charger between 6 and 16A in order to charge from 1.38kW to 3.68kW (1 phase). My goe-charger is an old one (v2), that cannot be switched between 1 and 3 phase charging. 

I'm using HACS SolarEdge https://github.com/binsentsu/home-assistant-solaredge-modbus to get actual values from my PV. Depending on the solar output, my automation switches between 6 and 16 A every couple of seconds. I'm fine with max. power of 3.68kW when charging at sunshine, since my PV has just 7kWp and the car is at home almost all day. 

Two phantastic custom components in HA (thanx to @cathiele and @bisentu) and a homegrown automation in one environment!

**Reply by jmtatsch** [reply]
How can you charge on one phase if you have a v2 charger that cannot switch between phases?
Can you post your automation?

**Reply by WolfgangSchmid** [reply]
> How can you charge on one phase if you have a v2 charger that cannot switch between phases?
> 

I have two different cables. 1-phase for home, 3-phase when I am away from home (or 11kW charge needed).

**Comment by morphZ** [comment]
I also started with the direct approach of setting the go-e charger to the free power (pv generation - house consumption), but quickly ran into some problems:
- go-e only supports setting the current, but the actual power depends also on the car. (I'm driving a kia and using the 3ph cable, I use a `reduced charge current` from the car's settings. This is also because I run the old hardware of go-e without the 3ph-1ph-switching)
- In winter month (like now, unfortunately :) there's not enough excess power to charge enough, so I needed a way to automatically use regular grid energy as well.

I came up with the following strategy, depending on the `soc` of the car:
- `soc` < 40% --> charge with max power
- 40% <= `soc` < 60% --> reduce charge power linearly from max to min
- `soc` >= 60% --> charge only with PV excess

To overcome the car dependency I implemented the PID-controller to bring the charge power to the desired value.

Does somebody now, how the strategy is with the mentioned ECVV?

**Comment by klacol** [comment]
My approach was, to run a python script in Home-Assistant on every change of the Solar-Edge values. Something [like this](https://gist.github.com/klacol/ca9bd7e18b2c5d060b16c542b2e7fd21).

My idea was, to have a script, that can apply any logic, that is needed and not to have a simple HA automation. I do not have it running in "production". I am seeking a good and stable solution.

**Comment by WolfgangSchmid** [comment]
Here is my approach. I've created a sensor to get the exceed in W ("verfuegbarfuerwb"), see my sensor.yaml below. 
This sensor value is used in 10 automations (from 6A to 16A) where the service goecharger.set_max_current is used to set the corresponding value depending on the available exceed. See the automation for 6A below.

[goeCharger_samples.txt](https://github.com/cathiele/homeassistant-goecharger/files/7739053/goeCharger_samples.txt)
